### PR TITLE
Calendar Bugfix: Calendar Global Usage despite having local Services

### PR DESCRIPTION
**Pull request by fhelfer**
Calendar Uses Global $DIC language even if Languages are set static to a different Language.
This language is not respected when Months are translated.

This is a huge issue for Certificates, because from the Certificate Component it's not possible to set a Date in the language of the Certificate Recipient

Best, @fhelfer

### CODE REVIEW

File: components/ILIAS/Calendar/classes/class.ilDatePresentation.php [hunk 1]
```diff
@@ -85,13 +99,15 @@ public static function resetToDefaults()
         self::setUseRelativeDates(true);
     }
 
-    public static function formatDate(ilDateTime $date, bool $a_skip_day = false, bool $a_include_wd = false, bool $include_seconds = false): string
-    {
+    public static function formatDate(
+        ilDateTime $date,
+        bool $a_skip_day = false,
+        bool $a_include_wd = false,
+        bool $include_seconds = false,
```
Comment (alex40724) [inline components/ILIAS/Calendar/classes/class.ilDatePresentation.php:86]: Why not adding an optional language user object and deriving the appropriate language object from that?

**Review by alex40724 — COMMENTED** [review]
Thanks for the PR. Would it be possible to pass an optional ilObjUser instance to formatDate() instead of using the static setters which mess with a "global state" like mechanism?

File: components/ILIAS/Calendar/classes/class.ilDatePresentation.php (hunk 1 above)
Comment (fhelfer) [reply components/ILIAS/Calendar/classes/class.ilDatePresentation.php:86]: The static Method setLanguage is already in use in many components. Some of which have not a clear direct formatDate() call. Without knowing the side effects, it's difficult to now change that.
I agree that passing this is a much better practice, but as it is already used, I cannot predict what effects that may have if we remove setLanguage(). And having both setLanguage and a parameter seems to open us up to race condition and confusion for developers.

Let me know your thoughts to this. Also, maybe start a discussion with other maintainers if that seems like a greater deal.

Best @fhelfer

**Comment by fhelfer** [comment]
> Thanks for the PR. Would it be possible to pass an optional ilObjUser instance to formatDate() instead of using the static setters which mess with a "global state" like mechanism?



> Thanks for the PR. Would it be possible to pass an optional ilObjUser instance to formatDate() instead of using the static setters which mess with a "global state" like mechanism?

As for passing a User, that would be not that much of a problem, but it felt like mixing different concepts. So maybe we wait the discussion on that.

**Comment by mjansenDatabay** [comment]
Hi @alex40724 and @smeyer-ilias ,

any further feedback or a merge of this PR would be highly appreciated, since six other PRs depend on this.

Best regards and thanks for caring,
Michael

File: components/ILIAS/Calendar/classes/class.ilDatePresentation.php [hunk 2]
```diff
@@ -154,59 +171,101 @@ public static function formatDate(ilDateTime $date, bool $a_skip_day = false, bo
...
+                            );
+                }
+            } else {
+                switch ($user->getTimeFormat()) {
+                    case ilCalendarSettings::TIME_FORMAT_24:
+                        return $date_str . $sep .
+                            $start->get(
+                                IL_CAL_FKT_DATE,
+                                'H:i',
+                                $user->getTimeZone()
```
Comment (alex40724) [inline components/ILIAS/Calendar/classes/class.ilDatePresentation.php:237]: @fhelfer Please have a look at these kind of changes. This costs really lot of time to check, if this code "has no changes at all". Please, if you provide bug fix PRs, do not include so many code style improvements.

**Review by alex40724 — APPROVED** [review]
Thanks. We checked the PR in detail. Please note that these kind of PRs are really hard to read, since it includes so many code style related changes. It would be better so separate only the bugfix related changes.

Please provide a separate PR for ILIAS 9 without the coding style changes.

**Pull request merged**